### PR TITLE
Comms builder for fully-async comms node

### LINKED ISSUE #1235: Allow protocol handlers to be registered in CommsBuilder

### PR DESCRIPTION
**Pull request by sdbondi**
## Description

Comms builder is the public interface for building a comms node. This PR
is an iteration on the previous builder which brings together all the new
async services.

It now handles hooking up the inbound and outbound "pipelines"
to the /tari/messaging protocol. (This used to be an external
responsibility). This interface is fully compatible with the existing pipeline (dht, tari_p2p etc)

It also allows custom protocols to be added.

An example initialization:

```rust
  let comms_node = CommsBuilder::new()
        .with_listener_address("/ip4/0.0.0.0/tcp/0".parse().unwrap())
        .with_transport(TcpTransport::default())
        .with_peer_storage(HashmapDatabase::new())
        .with_messaging_pipeline(
            pipeline::Builder::new()
                // Outbound messages will be forwarded "as is" to outbound messaging
                .with_outbound_pipeline(outbound_rx, std::convert::identity)
               // Up to 10 inbound messages concurrently processed by the given pipeline 
                .max_concurrent_inbound_tasks(10)
                // Inbound messages will be forwarded "as is" to inbound_tx
                .with_inbound_pipeline(SinkService::new(inbound_tx))
                .finish(),
        )
        .with_node_identity(node_identity)
        .with_protocols(
           Protocols::new()
             // Add protocol ids and notifiers (mpsc::Sender). Notifications are sent when a peer wants to speak a particular protocol
               .add(&[MY_CUSTOM_PROTOCOL_v1, MY_CUSTOM_PROTOCOL_v2], my_custom_notif)
               .add(&[DHT_PROTOCOL], dht_notif)
               .add(&[MEMPOOL_PROTOCOL], mempool_notif)
        )
        .spawn()
        .await
        .unwrap();
```

## Motivation and Context

Closes #1235 

## How Has This Been Tested?

New tests testing the custom protocol and messaging protocol setup with the new comms builder.

## Types of changes

* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [x] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [x] New Tests
* [ ] Documentation

## Checklist:

* [x] I'm merging against the `development` branch
* [x] I ran `cargo-fmt --all` before pushing
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.

### CODE REVIEW

File: comms/src/builder/builder_next.rs [hunk 1]
```diff
@@ -0,0 +1,446 @@
...
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+//  INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+//  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+//  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+//  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+//  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use crate::{
```
Comment (sdbondi) [inline comms/src/builder/builder_next.rs]: 👀

File: comms/src/builder/tests.rs [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+    types::CommsSubstream,
+};
+use bytes::Bytes;
+use futures::{channel::mpsc, AsyncReadExt, AsyncWriteExt, SinkExt, StreamExt};
+use std::{convert::identity, sync::Arc, time::Duration};
+use tari_storage::HashmapDatabase;
+use tari_test_utils::{collect_stream, unpack_enum};
+use tokio::runtime;
+
+async fn spawn_node(
```
Comment (sdbondi) [inline comms/src/builder/tests.rs]: 👀

File: comms/src/pipeline/builder.rs [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use crate::{message::InboundMessage, outbound_message_service::OutboundMessage, pipeline::SinkService};
```
Comment (sdbondi) [inline comms/src/pipeline/builder.rs]: 👀

File: comms/src/pipeline/inbound.rs [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+// INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+// USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use crate::bounded_executor::BoundedExecutor;
```
Comment (sdbondi) [inline comms/src/pipeline/inbound.rs]: 👀

File: comms/src/pipeline/outbound.rs [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+};
+use futures::{channel::mpsc, future, future::Either, stream::FusedStream, SinkExt, Stream, StreamExt};
+use log::*;
+use std::fmt::Debug;
+use tokio::runtime;
+use tower::{Service, ServiceExt};
+
+const LOG_TARGET: &str = "comms::pipeline::outbound";
+
+pub struct Outbound<TPipeline, TStream> {
```
Comment (sdbondi) [inline comms/src/pipeline/outbound.rs]: 👀

File: comms/src/builder/builder_next.rs [hunk 6]
```diff
@@ -0,0 +1,446 @@
...
+        self.dial_backoff = Some(Box::new(backoff));
+        self
+    }
+
+    pub fn with_transport<T>(self, transport: T) -> CommsBuilder<T, TInPipe, TOutPipe, TOutReq>
+    where
+        T: Transport + Unpin + Send + Sync + Clone + 'static,
+        T::Output: AsyncRead + AsyncWrite + Send + Sync + Unpin + 'static,
+    {
+        CommsBuilder {
```
Comment (philipr-za) [inline comms/src/builder/builder_next.rs]: Can simplify as:
```Rust
CommsBuilder {
    transport: Some(transport),
    ..self
}
```

File: comms/src/builder/builder_next.rs [hunk 7]
```diff
@@ -0,0 +1,446 @@
...
+    ) -> CommsBuilder<TTransport, I, O, R>
+    where
+        O: Service<R, Response = ()> + Clone + Send + 'static,
+        O::Error: fmt::Debug + Send,
+        O::Future: Send + 'static,
+        I: Service<InboundMessage> + Clone + Send + 'static,
+        I::Error: fmt::Debug + Send,
+        I::Future: Send + 'static,
+    {
+        CommsBuilder {
```
Comment (philipr-za) [inline comms/src/builder/builder_next.rs]: Can simplify as above

File: comms/src/builder/builder_next.rs [hunk 8]
```diff
@@ -0,0 +1,446 @@
...
+    messaging: MessagingProtocol,
+    messaging_event_tx: messaging::MessagingEventSender,
+    inbound_message_rx: mpsc::Receiver<InboundMessage>,
+
+    messaging_request_tx: mpsc::Sender<messaging::MessagingRequest>,
+    shutdown: Shutdown,
+    peer_manager: Arc<PeerManager>,
+}
+
+pub struct CommsNode {
```
Comment (philipr-za) [inline comms/src/builder/builder_next.rs]: Nit: Maybe just a comment that this is essentially a handle?

File: comms/src/builder/builder_next.rs (hunk 6 above)
Comment (sdbondi) [reply comms/src/builder/builder_next.rs]: I wish, but unfortunately you can't - `self` has to be the same type to use that syntax but in this case it isn't - the value we're returning is of a different type. There have been proposals in rust to fix this but nothing has made it into the language as yet.

File: comms/src/builder/builder_next.rs (hunk 7 above)
Comment (sdbondi) [reply comms/src/builder/builder_next.rs]: As above

File: comms/src/builder/builder_next.rs (hunk 8 above)
Comment (sdbondi) [reply comms/src/builder/builder_next.rs]: I'll add some comments

**Pull request merged**